### ISSUE TITLE
Eliminate Heap Allocations in Kernel

### ISSUE THREAD

**Issue by mmstick** [redox-os/redox#789]
I noticed that there's a lot of usage of collections that dynamically allocate on the heap inside the kernel, such as [this](https://github.com/redox-os/redox/blob/6927a4c5cfef4b97a58ba63b4f9d6d01dd8e9824/kernel/scheme/sys/memory.rs). It would be better to pass around stack-allocated arrays or at least take advantage of [stack-allocated vectors](https://docs.rs/arrayvec/0.3.20/arrayvec/). Kernel code should probably attempt to avoid heap allocations entirely.

**Comment by jackpot51** [comment]
Good idea. Kernel code is very stack limited, 64 KB by default. There are probably many cases where stack allocation could be used, with the proper understanding of its limitations.

**Comment by ticki** [comment]
Stack extension is a possible workaround the stack size problems.

**Comment by ticki** [comment]
(basically, set up guard pages, and "extend" the stack through new pages when its reached)

**Comment by jackpot51** [comment]
But is it really necessary on the kernel stack?

**Comment by ticki** [comment]
It'd be a huge problem if I could crash the kernel by giving some syscall parameters.

**Comment by jackpot51** [comment]
Give me an example of being able to, from system calls, exceed the kernel stack. I believe the only possible case is scheme call recursion

**Comment by ticki** [comment]
If we're going to allocate things like strings on the stack, you can input an arbitrary long string and overflow the stack.

**Comment by jackpot51** [comment]
We don't do that right now. I would be testing lengths if it were ever to be used

**Comment by ticki** [comment]
> We don't do that right now.

I know, but look at the title of this issue :P

**Comment by jackpot51** [comment]
In the case @mmstick mentions, for example, stack allocation cannot be used as the Vec is held after the open and must be available for read operations after the kernel stack has been completely unwinded by returning to userspace

**Comment by jackpot51** [comment]
Also, the case I mentioned of scheme call recursion will not overload the kernel stack either, it will just block the process. If the file scheme handler tries to open a file, for example.

**Comment by StefanoD** [comment]
If I remember correctly, Tannenbaum once said, that Minix 3 is completely(!) stack allocated. Maybe you could look there how they solve their problems.

**Comment by StefanoD** [comment]
Minix doesn't do anything dynamic in the kernel:

> My initial decision back in 1984 to have fixed-size messages throughout the system and avoid dynamic memory allocation (such as malloc) and a heap in the kernel has not been a problem and avoids problems that occur with dynamic storage management (such as memory leaks and buffer overruns).

http://cacm.acm.org/magazines/2016/3/198874-lessons-learned-from-30-years-of-minix/fulltext

Maybe you could look at the source code how they solve their problems.

**Comment by ksf** [comment]
As another reference: SeL4 requires callers of syscalls to provide it with the necessary heap space by passing in a suitable capability alongside. It has a very small, static, heap for own use, the rest is per-context. Those capabilities are revocable, doing that for a grant for a context spawn is a rather drastic way to say `kill -9`.
